awkward._nplikes.typetracer.TypeTracerArray._new needs further optimization, it is hot code

### Version of Awkward Array

2.6.1

### Description and code to reproduce

The present code is below:
```python3
    @classmethod
    def _new(
        cls,
        dtype: DType,
        shape: tuple[ShapeItem, ...],
        form_key: str | None = None,
        report: TypeTracerReport | None = None,
    ):
        self = super().__new__(cls)
        self.form_key = form_key
        self.report = report

        if not isinstance(shape, tuple):
            raise TypeError("typetracer shape must be a tuple")
        if not all(isinstance(x, int) or x is unknown_length for x in shape):
            raise TypeError("typetracer shape must be integers or unknown-length")
        if not isinstance(dtype, np.dtype):
            raise TypeError("typetracer dtype must be an instance of np.dtype")
        self._shape = shape
        self._dtype = dtype

        return self
```

in particular:
```python3
        if not isinstance(shape, tuple):
            raise TypeError("typetracer shape must be a tuple")
        if not all(isinstance(x, int) or x is unknown_length for x in shape):
            raise TypeError("typetracer shape must be integers or unknown-length")
        if not isinstance(dtype, np.dtype):
            raise TypeError("typetracer dtype must be an instance of np.dtype")
```
is quite expensive at the scale of thousands of awkward array kernel calls and in normal, non-development, operation none of these if statements ever evaluate to True. They are relatively expensive to evaluate when there is no data processing, see for example typetracing in dask-awkward.

We know already that commenting out these lines brings significant performance improvements.
A reasonable alternative to that would be wrapping this code in an if block.

```python3
    @classmethod
    def _new(
        cls,
        dtype: DType,
        shape: tuple[ShapeItem, ...],
        form_key: str | None = None,
        report: TypeTracerReport | None = None,
    ):
        self = super().__new__(cls)
        self.form_key = form_key
        self.report = report

        if __development_build__:
            if not isinstance(shape, tuple):
                raise TypeError("typetracer shape must be a tuple")
            if not all(isinstance(x, int) or x is unknown_length for x in shape):
                raise TypeError("typetracer shape must be integers or unknown-length")
            if not isinstance(dtype, np.dtype):
                raise TypeError("typetracer dtype must be an instance of np.dtype")
        self._shape = shape
        self._dtype = dtype

        return self
```

Where `__development_build__` is just a stand in name. I think the only requirements would be that the parameter to the if statement is a simple bool, no lookups (etc.), as this is known to be pretty hot code.

Savings are O(1s) for a complex HEP analysis when building a single few-thousand layer task graph. This corresponds to many minutes saved in processing for full-scale analysis.

@jpivarski @agoose77 @martindurant

The first check would be to determine if it _can_ have `__slots__`. If you declare something with `__slots__` and then try to assign to an unrecognized attribute, it will raise an error (in PYthon 3.10, anyway):

```python
 >>> class Something:
...     __slots__ = ["one", "two"]
...     def __init__(self):
...         self.one = 1
...         self.two = 2
... 
>>> something = Something()
>>> something.three = 3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Something' object has no attribute 'three'
```

So let's try giving it `__slots__` and seeing if any of the tests fail with an `AttributeError`.

```diff
index 1e2812be..58d21b7f 100644
--- a/src/awkward/_nplikes/typetracer.py
+++ b/src/awkward/_nplikes/typetracer.py
@@ -161,6 +161,7 @@ class TypeTracerReport:
 
 
 class TypeTracerArray(NDArrayOperatorsMixin, ArrayLike):
+    __slots__ = ("_form_key", "_report", "_shape", "_dtype")
     _dtype: numpy.dtype
     _shape: tuple[ShapeItem, ...]
```

--------------------

<p align="center">♫♫ (on-hold music) ♫♫</p>

--------------------

All of the awkward, dask-awkward tests pass.

Coffea tests have these failures on my computer:

```
tests/test_dataset_tools.py::test_apply_to_fileset_hinted_form FAILED
tests/test_dataset_tools.py::test_preprocess_calculate_form FAILED
=== 2 failed, 224 passed, 3 skipped, 1 xfailed, 297 warnings in 875.36s (0:14:35) ===
```

(14 minutes!) But they're both

```
            if save_form:
                form_str = uproot.dask(
                    tree,
                    ak_add_doc=True,
                    filter_name=no_filter,
                    filter_typename=no_filter,
                    filter_branch=partial(_remove_not_interpretable, emit_warning=False),
                ).layout.form.to_json()
>               dask.base.function_cache.popitem()
E               AttributeError: module 'dask.base' has no attribute 'function_cache'

src/coffea/dataset_tools/preprocess.py:89: AttributeError
```

(I have Dask 2024.2.1). It doesn't seem to be related.

So we _can_ add `__slots__` to TypeTracerArray if it makes a performance difference.

--- a/src/awkward/_nplikes/typetracer.py
+++ b/src/awkward/_nplikes/typetracer.py
@@ -164,6 +164,8 @@ class TypeTracerArray(NDArrayOperatorsMixin, ArrayLike):
     _dtype: numpy.dtype
     _shape: tuple[ShapeItem, ...]
 
+    runtime_typechecks = True
+
     def __new__(cls, *args, **kwargs):
         raise TypeError(
             "internal_error: the `TypeTracer` nplike's `TypeTracerArray` object should never be directly instantiated"
@@ -185,12 +187,13 @@ class TypeTracerArray(NDArrayOperatorsMixin, ArrayLike):
         self.form_key = form_key
         self.report = report
 
-        if not isinstance(shape, tuple):
-            raise TypeError("typetracer shape must be a tuple")
-        if not all(isinstance(x, int) or x is unknown_length for x in shape):
-            raise TypeError("typetracer shape must be integers or unknown-length")
-        if not isinstance(dtype, np.dtype):
-            raise TypeError("typetracer dtype must be an instance of np.dtype")
+        if cls.runtime_typechecks:
+            if not isinstance(shape, tuple):
+                raise TypeError("typetracer shape must be a tuple")
+            if not all(isinstance(x, int) or x is unknown_length for x in shape):
+                raise TypeError("typetracer shape must be integers or unknown-length")
+            if not isinstance(dtype, np.dtype):
+                raise TypeError("typetracer dtype must be an instance of np.dtype")
         self._shape = shape
         self._dtype = dtype
 

